src/common/test-jsonfds.c

```diff
@@ -0,0 +1,905 @@
+
+/* --- testing of printing --- */
+
+typedef struct
+{
+  char buffer[1024];
+  FILE *stream;
+
+  char expected_buffer[1024];
+  gchar *expected;
+
+  size_t pagesize;
+  char *accessible;
+  char *inaccessible;
+} Helpers;
+
+static void
+helpers_init (Helpers *helpers)
+{
+  /* set up the stream */
+  helpers->stream = fmemopen (helpers->buffer, sizeof helpers->buffer, "w");
+  setbuf (helpers->stream, NULL);
+  rewind (helpers->stream);
+
+  /* create a range of accessible bytes surrounded by memory that will
+   * cause a crash if accessed.
+   */
+  helpers->pagesize = sysconf(_SC_PAGE_SIZE);
+
+  /* allocate 3 pages of memory that will crash when accessed (PROT_NONE) */
+  char *region = mmap (NULL, 3 * helpers->pagesize, PROT_NONE,
+                       MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+  g_assert (region != MAP_FAILED);
+
+  /* punch a read/writable hole in the middle of the 3 pages */
+  int r = mprotect (region + helpers->pagesize, helpers->pagesize, PROT_READ | PROT_WRITE);
+  g_assert (r == 0);
+
+  /* record the first and last(+1) accessible bytes.  accessing the
+   * bytes immediately outside of this range is guaranteed to crash.
+   * this allows us to ensure proper memory behaviour of the code we're
+   * testing.
+   */
+  helpers->accessible = region + helpers->pagesize;
+  helpers->inaccessible = helpers->accessible + helpers->pagesize;
+
+  /* setup the expected buffer */
+  helpers->expected = helpers->expected_buffer;
+}
+
+static void
+helpers_clear (Helpers *helpers)
+{
+  char *region = helpers->accessible - helpers->pagesize;
+  munmap (region, 3 * helpers->pagesize);
+  helpers->accessible = NULL;
+  helpers->inaccessible = NULL;
+
+  helpers->expected = NULL;
+
+  fclose (helpers->stream);
+  helpers->stream = NULL;
+}
+
+static void
+helpers_expected (Helpers     *helpers,
+                  const gchar *expected)
+{
+  helpers->expected = stpcpy (helpers->expected, expected);
+}
+
+static void
+helpers_compare_expected (Helpers *helpers)
+{
+#if 0
+  g_debug ("\n\n%.*s\n\nVS\n\n%.*s\n\n",
+           (int) ftell (helpers->stream), helpers->buffer,
+           (int) (helpers->expected - helpers->expected_buffer), helpers->expected_buffer);
+#endif
+
+  g_assert_cmpint (ftell (helpers->stream), ==, helpers->expected - helpers->expected_buffer);
+  g_assert (memcmp (helpers->buffer, helpers->expected_buffer, ftell (helpers->stream)) == 0);
+}
+
+/* These could easily be a bit higher, but the running time explodes
+ * pretty quickly, and this test benefits from being run under valgrind.
+ */
+#define MIN_GROUPS  1
+#define MAX_GROUPS  3
+#define MAX_REPS    5
+
+static guint
+divmod (guint *residual,
+        guint  divisor)
+{
+  guint result = *residual % divisor;
+
+  *residual /= divisor;
+
+  return result;
+}
+
+static void
+test_string (Helpers    *helpers,
+             const char *key,
+             const char *value,
+             gint        length,
+             gint        min_length)
+{
```

Comment:
Suggestion: add an assertion that `strlen(value) > length`, so that the test below is actually meaningful.

Reply:
`length` is passed to be taken as the actual length of `value` (which is otherwise not nul-terminated)